### DISCUSSION TITLE
New Repository Structure

### DISCUSSION THREAD

**Discussion by tisoft** [SAIC-iSmart-API/documentation#1] [General]
As proposed here https://github.com/ReverseEngineeringDE/SAIC-API-Documentation/issues/64 we are moving the project contents to this new organization.

Anything related to that move can be discussed here.

**Comment by tisoft** [comment]
@tosate I have moved your python mqtt stuff here: 

https://github.com/SAIC-iSmart-API/saic-python-mqtt-gateway

I changed the docker push to use the github repo for now, since I obliviously don't have your docker password. 😄 

You should have admin rights for that repo.

**Comment by tosate** [comment]
Thank you @tisoft

It has been a busy week. I will look into it on the weekend.

**Comment by Tonno87** [comment]
Should we create new add-ons for Home Assistant? At the moment it's very difficult to add or change some things on the original repository... The pull request for the MQTT Gateway add-on is online since 9 days, but no reaction from "ReverseEngineeringDE".

**Reply by tisoft** [reply]
Sure, I think you should have the rights to create a new repo in this organization.

I would suggest `saic-home-assistant-addon` as project name.

**Reply by Tonno87** [reply]
Okay, i will give it a try. I think it's easier to use your docker hub images instead of jar or python files. But to do that you should add tags on it to allow updates with Home Assistant.

**Reply by tisoft** [reply]
I did a release about an hour ago:

`saicismartapi/saic-java-mqtt-gateway:0.1.0`

**Reply by Tonno87** [reply]
Very good 😀

**Reply by tosate** [reply]
I am running my home assistant instance in a docker container. Therefore, I think that I cannot install a home assistant add-on that also needs to run in a docker container.
To simplify things, I want to create a home assistant integration. Unfortunately, this seems to be very complicated and requires approval from the home assistant maintainers.

I have created the saic-ismart-client Python package and it is published on the [Python Package Index](https://pypi.org/project/saic-ismart-client/). Could this be used by the home assistant add-on? Maybe that would be a better solution.
What do you think @Tonno87?

**Reply by Tonno87** [reply]
Yes, add-ons are only working with HASS.IO, but HA in docker is slightly faster. HA-addons based on docker-images are easier to update because the users doesn't need to rebuild it manually. 

But i will create two add-ons, one with java and another with the python package to test both with HA.

**Reply by tosate** [reply]
I you have any questions or suggestions for improvements don't hesitate to contact me!

**Reply by lorddoener** [reply]
@Tonno87 is there any progress ? I would like to use the new HomeAssistant addon with MQTT support

**Reply by Tonno87** [reply]
At the moment: no progress... I didn't had much time last two weeks (family things). Maybe i can copy and paste my last pull request for this project to reduce your waiting time.

**Reply by lorddoener** [reply]
Take your time, I just wanted to express my interest ;-) I tried to install from your repo the Homeassistant addon in the branch MQTT gateway. However, I did not manage to start the MQTT gateway successfully in HomeAssistant.

Thanks for your time and effort!

**Reply by Tonno87** [reply]
Yes this is not possible because the uri points to the original GitHub repository :)
You can download the new folder and store it in your home assistant directory at /addons

**Reply by Tonno87** [reply]
> Take your time, I just wanted to express my interest ;-) I tried to install from your repo the Homeassistant addon in the branch MQTT gateway. However, I did not manage to start the MQTT gateway successfully in HomeAssistant.
> 
> Thanks for your time and effort!

The addon is online :)

**Reply by lorddoener** [reply]
Thank you for the great work. 

However, the addon seems to retrieve data permanently, but the car is currently not charged.

That should lower the 12V battery or ? I have over 30 updates within 1 minute. Is this normal?

**Reply by Tonno87** [reply]
Hm... Normally this addon works like a docker compose, but in the past i didn't use the mqtt gateway. So actually i can't say it's normal or not.

**Reply by Tonno87** [reply]
Could it be the "drivetrain/hvBatteryActive" setting? This topic is new in 0.2.0: "allow setting the drivetrain/hvBatteryActive/set state to force updates". default seems to be "false"

**Reply by Tonno87** [reply]
Tested the python gateway today... Works way better! It does nothing after the first start unless you charge or force an update 👍 But there's another problem... The latest tag seems not to force an update on Home Assistant, but i will try to fix it with an manual versioning.

**Reply by lorddoener** [reply]
> Could it be the "drivetrain/hvBatteryActive" setting? This topic is new in 0.2.0: "allow setting the drivetrain/hvBatteryActive/set state to force updates". default seems to be "false"

hvBatteryActive = false, but refresh/lastVehicleState and refresh/lastChargeState updates every second or have I misunderstood something?

**Reply by mschmiedel** [reply]
I added a MR to fix this on the java client, at some point it should behave closer to the Python Gateway... ;)

**Comment by tisoft** [comment]
I have closed all issues in the previous repository https://github.com/ReverseEngineeringDE/SAIC-API-Documentation and replaced the readme with a link to this organization.

**Comment by mschmiedel** [comment]
I see that some of the configurations that someone would like to have for the gateway (refreshPeriodActive, refreshPeriodInactive, refreshMode), and maybe some status messages might be interesting to have as retained messages (which some already are).

However, it would be great if the gateways would start by reading all the retained messages from mqtt before resetting default values, so that:

1. Configurations that someone executed are kept
2. Status of the vehicle is also kept and the gateway goes on where it was left before shutdown

What do you think? If so, should we agree on what topics should be tried to be read before setting default values?

**Reply by tisoft** [reply]
I would keep all `/set` topics not-retained, since I consider them as command, not as state.

I agree that it would be nice to read the current configuration and vehicle state on restart, to continue where it left off.

So my proposal would be to read the already retained base topics (the part before the `/set`) on start of the gateway. If they are found use the values from them, if not fallback to default values.

I would read the following topics:
```
drivetrain/charging (might be unnecessary, if we keep the behavior of polling at start)
drivetrain/running (might be unnecessary, if we keep the behavior of polling at start)
drivetrain/remoteClimateState (might be unnecessary, if we keep the behavior of polling at start)
drivetrain/backWindowHeat (might be unnecessary, if we keep the behavior of polling at start)
refresh/mode
refresh/period/active
refresh/period/inActive
refresh/period/afterShutdown (name still TBD)
```

What do you think?

**Reply by mschmiedel** [reply]
Yep, the states that can be read from the car are rather unnecessary, maybe I would propose the lastCarActivity timestamp and other similar to avoid that a crashing gateway bombs the 12V battery... 😉

Most important are the refresh configurations that are meant to configure the gateway rather than the car...
